apps/twofactor_backupcodes/lib/Service/BackupCodeStorage.php

```
@@ -68,10 +73,28 @@ public function createCodes(IUser $user, $number = 10) {
 			array_push($result, $code);
 		}
 
+		$this->publishEvent($user, 'codes_generated');
+
 		return $result;
 	}
 
 	/**
+	 * Push an event the user's activity stream
+	 *
+	 * @param IUser $user
+	 * @param string $event
+	 */
+	private function publishEvent(IUser $user, $event) {
+		$activity = $this->activityManager->generateEvent();
+		$activity->setApp('twofactor_backupcodes')
+			->setType('twofactor')
+			->setAuthor($user->getUID())
+			->setAffectedUser($user->getUID());
+		$activity->setSubject($event);
+		$this->activityManager->publish($activity);
```

which one? The API doc does not mention any specific exception. Catching \Exception is bad IMO because it potentially swallows unrelated stuff and bugs go unnoticed :-/